File: Behavioral/Lab20/Typescript/src/Entities/States/sellable_state.ts
import { Customer } from "../customer";
import { RealEstate } from "../real_estate";
import { EstateState } from "./estate_state";

export class SellableState extends EstateState
{
    public override rent(customer: Customer, startDate: Date, endDate: Date): void {
        console.log(`Customer ${customer.nameAndSurname} can't rent an estate`);
    }

    public override sell(customer: Customer): void {
        if(!this._estate.isAlreadySold){
            this._estate.isAlreadySold = true;
            console.log(`Customer ${customer.nameAndSurname} sold an estate`);
        }
        else
            console.log(`Customer ${customer.nameAndSurname} can't sell an estate because it is already sold`);
    }

    public override setEstateToBuy(customer: Customer, price: number): void {
        console.log(`Customer ${customer.nameAndSurname} can't buy an estate`);
    }

}

File: Behavioral/Lab20/C#/Entities/States/SellableState.cs
using System;
using System.Collections.Generic;
using System.Text;
using C_.Interfaces;

namespace C_.Entities.States
{
    public class SellableState : EstateState
    {
        public SellableState(RealEstate estate) : base(estate)
        {
            
        }
        public override void Rent(Customer customer, DateTime startDate, DateTime endDate)
        {
            Console.WriteLine($"Customer {customer.NameAndSurname} can't rent an estate");
        }

        public override void Sell(Customer customer)
        {
            if(!_estate.IsAlreadySold){
                _estate.IsAlreadySold = true;
                Console.WriteLine($"Customer {customer.NameAndSurname} sold an estate");
            }
            else
                Console.WriteLine($"Customer {customer.NameAndSurname} can't sell an estate because it is already sold");
        }

        public override void SetEstateToBuy(Customer customer, double price)
        {
            Console.WriteLine($"Customer {customer.NameAndSurname} can't buy an estate");
        }

    }
}
